### DISCUSSION TITLE
How to implement alternative script syntaxes

### DISCUSSION THREAD

**Discussion by aramallo** [cozodb/cozo#201] [Q&A]
Hi, I am evaluating Cozo for several projects and I am trully impressed by it so far, so thanks for the hard work in making this an OSS project.

I have a background in Datalog (having implemented a distributed semantic graph store with a distributed datalog evaluator) and I am curious about some key Cozo syntax decisions and wondering how I could go about implementing an alternative syntax.

I am thinking about two alternatives:
1. Implement a new syntax in Rust (forking Cozo)
2. Implement a side library provided Cozo can expose a `run_ast` function which takes a valid Cozo AST generated externally (by my syntax parser).

In addition, a couple of key questions regarding some syntax decisions:

* **The star `*` prefix for extensional relations (stored relations).** Why does Cozo need the user to distinguish between internsional (constant rules) and extensional relations given that Cozo evidently has a system catalogue containing all extensional relations? 
* **The need for `<-` and `<~`**. Any particular reason the parser will need this hint, as opposed to recognising the body contains a Fixed Rule atom?

The above questions are not intended as critique of the syntax but to understand how I would need to go about adding support for a different syntax that would allow me to migrate or integrate Cozo into my existing system.

I am new to Rust too, so any pointers on how I could go about implementing the above would be trully appreciated.

**Comment by aramallo** [comment]
I am obvisouly not considering the option of transpiling my syntax into Cozo Script as it will add complexity and latency. I think the second idea mentioned above is a better option than transpiling.

**Comment by gisborne** [comment]
Absolutely +1 to this.
